staging/src/k8s.io/client-go/ARCHITECTURE.md

```diff
@@ -0,0 +1,161 @@
+response processing until a method like `.Into(&pod)` is called. This separation is key to
+supporting different client models from a common base.
+
+### Endpoint Interactions
+
+*   **Content Negotiation:** The client uses HTTP `Accept` headers to negotiate the wire format
+    (JSON or Protobuf). A key performance optimization using this mechanism is the ability to
+    request metadata-only objects via the `as=PartialObjectMetadata;g=meta.k8s.io;v=v1` Accept custom parameter.
+    Also the `as=Table;g=meta.k8s.io;v=v1` Accept custom parameters may be used to request lists as tables. 
+*   **Subresources:** The client can target standard subresources like `/status` or `/scale` for
+    object mutations, and it can also handle action-oriented subresources like `/logs` or
+    `/exec`, which often involve streaming data.
+*   **List Pagination:** For `LIST` requests, the client can specify a `limit`. The server will
+    return up to that many items and, if more exist, a `continue` token. The client is
+    responsible for passing this token in a subsequent request to retrieve the next page.
+    Higher-level tools like the `Reflector`'s `ListerWatcher` handle this logic automatically.
+*   **Streaming Watches:** A `WATCH` request returns a `watch.Interface` (from
+    `k8s.io/apimachinery/pkg/watch`), which provides a channel of structured `watch.Event`
+    objects (`ADDED`, `MODIFIED`, `DELETED`, `BOOKMARK`). This decouples the watch consumer from
+    the underlying streaming protocol.
+
+### Errors, Warnings, and Rate Limiting
+
+*   **Structured Errors:** The client deserializes non-2xx responses into a structured
+    `errors.StatusError`, enabling programmatic error handling (e.g., `errors.IsNotFound(err)`).
+*   **Warnings:** It processes non-fatal `Warning` headers from the API server via a
+    `WarningHandler`.
+*   **Client-Side Rate Limiting:** The `QPS` and `Burst` settings in `rest.Config` are the
+    client's half of the contract with the server's API Priority and Fairness system.
+*   **Server-Side Throttling:** The client's default transport automatically handles HTTP `429`
+    responses by reading the `Retry-After` header, waiting, and retrying the request.
+
+## Typed and Dynamic Clients
+
+To handle the extensible nature of the Kubernetes API, `client-go` provides two primary client
+models.
+
+The **`kubernetes.Clientset`** provides compile-time, type-safe access to core, built-in APIs.
+
+The **`dynamic.DynamicClient`** represents all objects as `unstructured.Unstructured`, allowing it
+to interact with any API resource, including CRDs. It relies on two discovery mechanisms:
+1.  The **`discovery.DiscoveryClient`** determines *what* resources exist. The
+    **`CachedDiscoveryClient`** is a critical optimization that caches this data on disk to solve
+    the severe N+1 request performance bottleneck that can occur during discovery.
```

Comment:
N+1 request is largely a nonissue after aggregated discovery, but disk cache is definitely a performance improvement and should be recommended.